Feature/#7 upload image

HTH016 opened the pull request:
## AWS S3 이미지 업로드 기능 추가 
### 회원가입 프로필 이미지 업로드 
#### 업로드를 하지 않으면 S3 bucket에 저장된 default image 적용
 
### post 작성 시 이미지 업로드 
#### 0개 ~ 20개 까지 허용

 S3 Service, S3Config
PostImage Entity, Repository

Multipart 파일 테스트를 위한 Converter 추가

이미지 업로드 기능 테스트 추가

HTH016 committed:
#07
AWS S3 연동
회원가입 프로필 업로드 추가
마이페이지 텍스트, 이미지파일 불러오기 확인차 구현

HTH016 committed:
Merge branch 'feature/#5-post' into feature/#7-upload-image

# Conflicts:
#	src/main/java/flab/Linkedlog/controller/MemberController.java
#	src/main/java/flab/Linkedlog/entity/Post.java
#	src/main/java/flab/Linkedlog/service/MemberService.java

HTH016 committed:
#07 (수정)
이전 branch 수정사항 반영
postImage entity, Repository 추가
post에 이미지 업로드 기능 구현중 임시 저장

HTH016 committed:
Merge branch 'feature/#5-post' into feature/#7-upload-image

# Conflicts:
#	src/main/java/flab/Linkedlog/service/PostService.java
#	src/test/java/flab/Linkedlog/postTest/CreatePostTests.java

HTH016 committed:
#07 (수정)
이전 branch 확인을 위한 임시 커밋

HTH016 committed:
#07 (수정)
post 이미지 업로드 테스트
회원가입 프로필 이미지 업로드 테스트

mason1155 commented on src/main/java/flab/Linkedlog/service/MemberService.java:
```diff
@@ -41,6 +48,14 @@ public Long signUp(SignUpRequest signUpDto) {
                 .build();
 
+        if (profileImage != null && !profileImage.isEmpty()) {
+            String imageKey = s3Service.uploadFile(profileImage);
```
트랜잭션 내부에 이미지 업로드하게 되면 DB 트랜잭션을 오래 가지고 있는 단점이 있습니다.
트랜잭션 범위 밖에서 이미지 업로드 하게 해주세요

mason1155 commented on src/main/java/flab/Linkedlog/service/PostService.java:
```diff
@@ -22,30 +27,51 @@ public class PostService {
+        }
+        if (images != null && images.size() > 20) {
+            throw new IllegalArgumentException();
         }
```
위 요청에 대한 검증은 Controller 레이어에서 진행해주세요

mason1155 commented on src/main/java/flab/Linkedlog/service/PostService.java:
```diff
@@ -22,30 +27,51 @@ public class PostService {
+            List<PostImage> postImages = new ArrayList<>();
+            for (MultipartFile image : images) {
+                if (!image.isEmpty()) {
+                    String imageUrl = s3Service.uploadFile(image);
```
여기도 이미지 업로드는 트랜잭션 범위 바깥에서 해주세요

mason1155 replied:
이미지 업로드가 오래 걸리는 작업인대 병렬로 하면 더 성능이 좋지 않을까요?

mason1155 commented on src/main/java/flab/Linkedlog/service/S3Service.java:
```diff
@@ -0,0 +1,61 @@
+
+
+    // 이미지 업로드
+    public String uploadFile(MultipartFile file) throws IOException {
```
버킷이랑 폴더 명을 매개변수로 받아서 확장성을 높이시죠
용도에 맞는 버킷 및 폴더 지정이 되게 해주세요

mason1155 commented on src/test/java/flab/Linkedlog/memberTest/SignUpIntegrationTests.java:
```diff
@@ -142,10 +181,77 @@ void signUpDuplicateTest() throws Exception {
+                        .file("profileImage", profileImage.getBytes())
+                        .file(signUpRequestPart)
+                        .contentType(MediaType.MULTIPART_FORM_DATA))
+                .andReturn();
```
MockMvc를 쓰면 응답에 대한 검증은 assertThat 보단 andExpect를 쓰는게 더 일반적이긴해요

HTH016 committed:
Merge branch 'feature/#5-post' into feature/#7-upload-image

# Conflicts:
#	src/main/java/flab/Linkedlog/controller/PostController.java
#	src/main/java/flab/Linkedlog/service/PostService.java

HTH016 committed:
#07 (수정)
S3Service 메서드 bucketName, folderName 을 argument에 추가, 기능에 따른 S3 버킷 폴더 분리
Async 설정, 이미지 업로드 Async로 트랜잭션 분리, 병렬처리 하도록 수정

mason1155 commented on src/main/java/flab/Linkedlog/service/PostService.java:
```diff
@@ -116,7 +146,18 @@ public PostDetailResponse getPostDetailById(Long categoryId, Long postId) {
     }
+
+    @Async
+    public CompletableFuture<PostImage> uploadPostImageAsync(MultipartFile image, Post post) throws IOException {
```
이렇게 한 클래스 내에서 @Async 메서드를 호출하면 @Async가 적용 안됩니다. 
원인은 찾아보시죠

mason1155 commented on src/main/java/flab/Linkedlog/service/MemberService.java:
```diff
@@ -76,6 +99,26 @@ public String login(LogInRequest logInRequest) {
+
+
+    @Async
+    public CompletableFuture<Void> uploadProfileImageAsync(MultipartFile profileImage, Member member) throws IOException {
```
여기도     @Async는 동작 안합니다.

mason1155 commented on src/main/java/flab/Linkedlog/service/PostService.java:
```diff
@@ -25,29 +33,46 @@ public class PostService {
+                    futureList.add(uploadFuture);
+                }
+            }
+            List<PostImage> postImages = futureList.stream()
```
커멘트 남겼듯이 createPost 메서드 내에서 이미지를 업로드하면 트랜잭션을 오랫동안 잡고 있게 됩니다.
이미지를 업로드하고 createPost 내에 업로드한 imageUrl들을 전달하게 수정해주시죠.

HTH016 committed:
#07
AsyncConfig 설정 변경
ImageUploadService로 async 메서드 분리
async로 이미지 업로드만 하도록 수정

mason1155 reviewed (approved):
고생하셨어요 👍 👍
